.github/workflows/test.yaml

```diff
@@ -1,77 +1,78 @@
-name: "test"
+name: "scaffolding-ci"
 
 on:
   # Trigger the workflow on push or pull request,
   # but only for the main branch
   push:
-    branches: [ develop, develop-0.1, develop-0.2 ]
+    branches: [develop, develop-0.1, develop-0.2]
   pull_request:
-    branches: [ develop, develop-0.1, develop-0.2 ]
+    branches: [develop, develop-0.1, develop-0.2]
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref_name }}
+  cancel-in-progress: true
 
 jobs:
-  test_fmt_lint:
+  cargo_test:
+    name: cargo_test
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v4
-
-      - name: Maximize build space
-        if: runner.os == 'Linux'
-        uses: AdityaGarg8/remove-unwanted-software@v2
+      - uses: dtolnay/rust-toolchain@stable
         with:
-          remove-dotnet: 'true'
-          remove-android: 'true'
-          remove-codeql: 'true'
-          remove-docker-images: 'true'
+          toolchain: 1.80.0
+      - uses: Swatinem/rust-cache@v2
+      - run: cargo test --no-fail-fast
 
-      - name: Install nix
-        uses: cachix/install-nix-action@v27
+  rustfmt:
+    name: rustfmt
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: dtolnay/rust-toolchain@stable
         with:
-          install_url: https://releases.nixos.org/nix/nix-2.24.4/install
+          toolchain: 1.80.0
+          components: rustfmt
+      - run: cargo fmt --all -- --check
 
-      - uses: cachix/cachix-action@v15
+  clippy:
```

Comment:
why separate these into different jobs? I'd do all Cargo work in one job as different steps. Otherwise you're cloning the repo and setting up the toolchain multiple times

Reply:
I think separating them is good for failure isolation, faster feedback and parallelization. The time saved from repo cloning and toolchain setup is quite minimal, especially with caching set up. What do you think?

Reply:
I'd probably make these steps under one job, but if you've looked at the run times and you're happy then this works fine